autopush/endpoint.py

```diff
@@ -447,39 +455,42 @@ def _uaid_lookup_results(self, result):
         if router_key == "simplepush":
             self.version, data = parse_request_params(self.request)
             self._client_info['message_id'] = self.version
+        elif router_key == "webpush":
+            data = self.request.body
+            if "ttl" not in self.request.headers:
+                ttl = None
+            # We need crypto headers for messages with payloads.
+            req_fields = ["content-encoding", "encryption"]
+            if data and not all([x in self.request.headers
+                                 for x in req_fields]):
+                log.msg("Client error", status_code=400, errno=101,
+                        **self._client_info)
+                return self._write_response(400, 101)
+            if ("encryption-key" in self.request.headers and
+                    "crypto-key" in self.request.headers):
+                log.msg("Client error", status_code=400, errno=110,
+                        **self._client_info)
+                return self._write_response(
+                    400, 110, message="Invalid crypto headers")
+            self._client_info["message_size"] = len(data) if data else 0
         else:
-
+            # GCM/APNS pull data out of body
             data = self.request.body
-            if router_key == "webpush":
-                if "ttl" not in self.request.headers:
-                    log.msg("Missing TTL Header",
-                            status_code=400,
-                            **self._client_info)
-                    self._write_response(status_code=400,
-                                         errno=111,
-                                         message="Missing TTL header")
-                    return
-                # We need crypto headers for messages with payloads.
-                req_fields = ["content-encoding", "encryption"]
-                if data and not all([x in self.request.headers
-                                     for x in req_fields]):
-                    log.msg("Missing Crypto headers", **self._client_info)
-                    return self._write_response(400, 101)
-                if ("encryption-key" in self.request.headers and
-                        "crypto-key" in self.request.headers):
-                    log.msg("Both encryption and crypto keys in headers",
-                            **self._client_info)
-                    return self._write_response(
-                        400, 110, message="Invalid crypto headers")
-                self._client_info["message_size"] = len(data) if data else 0
 
-        try:
-            ttl = int(self.request.headers.get("ttl", "0"))
+        if "ttl" not in self.request.headers:
+            ttl = None
+        elif VALID_TTL.match(self.request.headers["ttl"]):
```

Comment:
Throwing an exception is quite a bit more expensive than a compiled regex match call. Exception building requires a complete set of trace-back frames and all the associated objects. PyPy can optimize this out of the way, but cpython doesn't.